[docs] Add human-in-the-loop policy for tool-assisted contributions

rnk opened the pull request:
Over the course of 2025, we observed an increase in the volume of LLM-assisted nuisance contributions to the project. Nuisance contributions have always been an issue for open-source projects, but until LLMs, we made do without a formal policy banning such contributions. However, LLMs are here, so we are adopting this policy, abbreviated as "human in the loop", which requires that every contribution has a human author attesting to the value of that contribution, and that it is high enough quality that it is worth the time it takes to review the contribution.

This policy evolved over time based on community input from the following Discourse threads and a few area team and LLVM project council meetings:
* [Our AI policy vs code of conduct and vs reality](https://discourse.llvm.org/t/our-ai-policy-vs-code-of-conduct-and-vs-reality/88300)
* [[RFC] LLVM AI tool policy: start small, no slop
](https://discourse.llvm.org/t/rfc-llvm-ai-tool-policy-start-small-no-slop/88476)
* [[RFC] LLVM AI tool policy: human in the loop](https://discourse.llvm.org/t/rfc-llvm-ai-tool-policy-human-in-the-loop/89159)

rnk committed:
[docs] Strengthen our quality standards and connect AI contribution policy to it

Open source projects are often the target of nuisance contributions
(think typo corrections), but prior to the advent of AI tools, we hadn't
had to take a strong policy stance discouraging these contributions.
However, as the tools have improved, the skill required to make
plausible contributions has decreased, and burdensome nuisance
contributions have become a problem for our project, so I think we
should adopt a stronger quality bar policy.

I've updated our existing Quality section to frame everything in terms
of a golden rule that contributions should pass subjective maintainer
quality standards, held in balance against our desire to welcome new
contributors.

I then connected the AI policy section to our core quality policy, to
try to make the AI policy section tool-neutral. My goal is to minimize
the distinction between traditional mass-refactoring tools, and the
costs of those migrations, and AI-assisted contributions.

This is following up on a discussion in the July 17 LLVM project council
meeting.

boomanaiden154 commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -1458,23 +1497,66 @@ permission.
+attentional resources of LLVM project maintainers, the onus is on contributors
+to justify why their contributions are not burdensome and exceed our `quality`_
+bar. Contributors who repeatedly send low-quality contributions to our project
+will be subject to escalating moderation actions and eventually a project ban.
```
Is there a formal process for how a ban would be decided?

I know in the past it has gone through the code of conduct committee. I think it's important to explicitly signal that these cases are decided through a fair process rather than just handed out without any investigation.

cmtice commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -1458,23 +1497,66 @@ permission.
+has increased the asymmetry between the work of producing a contribution, and
+the work of reviewing the contribution. In order to protect the time and
+attentional resources of LLVM project maintainers, the onus is on contributors
+to justify why their contributions are not burdensome and exceed our `quality`_
```
Do they need to exceed it or just meet it?

rnk committed:
Updating text to center it around extractive/non-extractive contributions

nikic commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -293,57 +293,71 @@ warranted when performing a code review.
+the LLVM project to be welcoming and open to aspiring compiler engineers who are
+willing to invest time and effort to learn and grow, because growing our
+contributor base and recruiting new maintainers helps sustain the project over
+the long term. We therefore label pull requests from new contributors and
```
We do? What's the label for that?

rnk replied:
I based this on glancing at this github workflow:
https://github.com/llvm/llvm-project/blob/main/.github/workflows/new-prs.yml#L24

On second glance, we don't have a label, but we do post a greeting. I'll just update the text to match reality, but I think a label would be helpful too, since it can be used to specifically focus on reviewing PRs from new contributors. I realize I'm speaking as someone who only reviews things sent to me, but this seems like a good aspirational goal.

nikic commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -424,6 +438,40 @@ squashing and merging PRs.
+We prefer for this to be handled before submission but understand that it isn't
+possible to test all of this for every submission.  Our build bots and nightly
+testing infrastructure normally finds these problems.  A good rule of thumb is
+to check the nightly testers for regressions the day after your change.  Build
```
What nightly testers?

rnk replied:
This is stale text, but back in the day, llvm-test-suite was run nightly, I believe, and results were accumulated in LNT. [Docs](https://llvm.org/docs/lnt/intro.html) make reference to "nightly" there. I'll link it to make it clearer, but surely this could use rewriting.

nikic reviewed (commented):
Some high level feedback:

 * I think this is a bit too verbose for a policy document. Some of this (like the references to other AI policies) are something to mention in an RFC to justify how we arrived here, but they are not something every person reading the finished policy needs to look at. Similarly, I'm not sure we should be including the meta philosophy around "extractive" contributions in the policy document. We should be able to explain this more concisely -- is the contribution worth the required investment of reviewer time?
 * One of the motivations behind revisiting this policy is to avoid wasting reviewer time on low-quality AI generated PRs. I think what I'm currently missing from this policy is a means by which a reviewer can reject a low-quality PR without significant justification. Basically, this needs to be at the level of rejecting a PR with a copy&paste comment pointing to the policy.

rnk reviewed (commented):
I had pending comments, sorry these are stale, I wrote them last week. It makes more sense for me to send these as one update, and then focus on new feedback post project council meeting.

rnk commented:
> Some high level feedback:
> 
> * I think this is a bit too verbose for a policy document. Some of this (like the references to other AI policies) are something to mention in an RFC to justify how we arrived here, but they are not something every person reading the finished policy needs to look at. Similarly, I'm not sure we should be including the meta philosophy around "extractive" contributions in the policy document. We should be able to explain this more concisely -- is the contribution worth the required investment of reviewer time?

My view is that these references are a good way of capturing the "alternatives considered", which is often the most important part of any technical design document. They are also likely to be helpful in capturing the "why" of any policy, which is something that the next policy doc editor will probably care about (I now find myself in this position). I've personally received a lot of links to what other projects are doing, and so the section serves as a way of saying "yes, we looked at all that, and decided we want to make compromise X".

I think the Developer Policy doc is just growing, and it needs to be split up (LangRef has grown similarly), and it also needs to be adapted to apply to the entire LLVM umbrella project. Much of the text is focused on LLVM the optimizer and code generator.

Re: extractive contributions, @chandlerc gave me the opposite feedback, that I should center the ideas of extractive and non-extractive contributions. I defer to the opinions of active maintainers, but that's why I went philosophical. Quality is inherently subjective and philosophical, at the end of the ady. 

> * One of the motivations behind revisiting this policy is to avoid wasting reviewer time on low-quality AI generated PRs. I think what I'm currently missing from this policy is a means by which a reviewer can reject a low-quality PR without significant justification. Basically, this needs to be at the level of rejecting a PR with a copy&paste comment pointing to the policy.

This is true, I will take action on that, I got the same feedback from Aaron. I do need to get that PR-copy-pasta message into the proposal near the top.

rnk committed:
Address feedback on nightly testing and add copy-pasted message

rnk committed:
Update wording: We comment on first-time PRs, we do not label them (yet)

rnk committed:
Merge branch 'main' into tool-policy

rnk committed:
Move all the text into a new policy doc

rnk committed:
Restore original Developer Policy text with minimal edits

rnk committed:
Merge branch 'main' into tool-policy

rnk committed:
Merge branch 'main' into tool-policy

rnk committed:
Incorporate the Fedora policy, indicate it carries a CCA license, retain attribution, and add a point about contribution size

rnk committed:
More minor edits

rnk committed:
Add objective guideline of 150 additional lines of code

maflcko reviewed (approved):
Makes sense to clarify the AI policy section and move it to a new doc. I'd say most of it was already common sense and covered by the previous section. However, it can't hurt to spell it out a bit more verbosely. Also, the new *Handling Violations* sections is useful for this project, going forward, on how to approach violations explicitly.

asb commented on llvm/docs/AIToolPolicy.md:
```diff
@@ -0,0 +1,154 @@
+## Handling Violations
+
+If a maintainer judges that a contribution is *extractive* (i.e. it is
+generated with tool-assistance or simply requires significant revision), they
```
I believe these lines don't capture the intent of the policy. It suggests any contribution generated with tool assistance is extractive. Perhaps "it is generated with tool-assistance in a way inconsistent with our AI tool usage policy"?

cmtice commented on llvm/docs/AIToolPolicy.md:
```diff
@@ -0,0 +1,154 @@
+  pull request description, commit message, or wherever authorship is normally
+  indicated for the work. For instance, use a commit message trailer like
+  Assisted-by: <name of code assistant>. This transparency helps the community
+  develop best practices and understand the role of these new tools.
```
Not sure if you would want to mention it, but it can also help build trust in usimg these tools. Maybe? ("This nice PR was assisted by such-and-so tool -- maybe it's not so bad for that tool to help with PRs...")

rnk committed:
Merge branch 'main' into tool-policy

rnk committed:
Try again, focusing on the idea of having a human in the loop.

Doc remainder will be edited, I'm mostly trying to push something out and defeat writer's block

rnk committed:
Merge branch 'main' into tool-policy

rnk committed:
December editting pass, make another attmept

hubert-reinterpretcast commented on llvm/docs/AIToolPolicy.md:
```diff
@@ -1,79 +1,37 @@
+and is fully accountable for their contributions. Contributors should be
+sufficiently confident that the contribution is high enough quality that asking
+for a review is a good use of scarce maintainer time, and they should be **able
+to answer quesions about their work** during review.
```
Typo:
```suggestion
to answer questions about their work** during review.
```

hubert-reinterpretcast commented on llvm/docs/AIToolPolicy.md:
```diff
@@ -0,0 +1,162 @@
+use of tools under this policy.
+
+This policy extends beyond code contributions and includes, but is not limited
+to, the following kinds of contributions:
```
Remove duplicate mention of code contributions.
```suggestion
This policy includes, but is not limited to, the following kinds of
contributions:
```

rnk replied:
done

rnk committed:
Wordsmith

rnk committed:
Move the labelling and transparency policy text earlier so it does not get lost

rnk committed:
Apply suggestions from code review

Co-authored-by: Hubert Tong <hubert.reinterpretcast@gmail.com>
Co-authored-by: maflcko <6399679+maflcko@users.noreply.github.com>

rnk committed:
Respond to comments

rnk committed:
Fix doc URL

asb commented on llvm/docs/AIToolPolicy.md:
```diff
@@ -0,0 +1,162 @@
+## Handling Violations
+
+If a maintainer judges that a contribution is *extractive* (i.e. it doesn't
+comply with this policy), they should copy-paste the following response to
```
Would `judges that a contribution is *extractive* and does not comply with this policy` capture this in a more straightforward way @hubert-reinterpretcast?

hubert-reinterpretcast replied:
> Would `judges that a contribution is *extractive* and does not comply with this policy` capture this in a more straightforward way @hubert-reinterpretcast?

Yes.

pranavk commented on llvm/docs/AIToolPolicy.md:
```diff
@@ -0,0 +1,166 @@
+LLVM's policy is that contributors can use whatever tools they would like to
+craft their contributions, but there must be a **human in the loop**.
+**Contributors must read and review all LLM-generated code or text before they
+ask other project members to review it.** The contributor is always the author
```
The QEMU AI policy, though much stricter, has section on [Exceptions](https://www.qemu.org/docs/master/devel/code-provenance.html#exceptions). Perhaps it may be a good idea to have something similar to be explicit that LLVM welcomes discussion on exceptions via the RFC process.

KyrylR commented on llvm/docs/AIToolPolicy.md:
```diff
@@ -0,0 +1,162 @@
+    This PR appears to be extractive, and requires additional justification for
+    why it is valuable enough to the project for us to review it. Please see
+    our developer policy on AI-generated contributions:
+    http://llvm.org/docs/AIToolPolicy.html
```
This link is broken

KyrylR replied:
Got it, thanks!

rnk committed:
Merge branch 'main' into tool-policy

Move the good-first-issue doc into the separate policy doc

rnk committed:
Iterate on violation handling text

rengolin reviewed (approved):
The text reflects my interpretation of consensus on the various RFCs in the forum, not necessarily my personal opinion. I believe this will make it easier to discuss other topics that have flourished since this subject was raised and will make it easier for reviewers to spend their times wisely. Thank you all for the long and hard work.

nikic reviewed (approved):
LGTM

rnk committed:
Final suggested tweak

rnk committed:
Merge branch 'main' into tool-policy
